src/nedap/speced/def/impl/analysis.cljc

```diff
@@ -109,6 +109,78 @@
                     args-with-proper-tag-hints (ensure-proper-type-hints clj? args)]
                 (apply list args-with-proper-tag-hints prepost body))))))
 
+(defn add-checking-form [tails ret-spec clj?] ;; deserves a test now
+  (->> tails
+       (map (fn [[args maybe-prepost & maybe-rest-of-body :as tail]]
+              (let [rest-of-body? (-> maybe-rest-of-body seq)
+                    body (if rest-of-body?
+                           (cons maybe-prepost rest-of-body?)
+                           [maybe-prepost])
+                    has-prepost? (and (not= [maybe-prepost] body)
+                                      (map? maybe-prepost))
+                    body (if has-prepost?
+                           (rest body)
+                           body)
+                    specs-from-destructurings (extract-specs-from-destructurings clj? args)
+                    non-destructured-args (->> args (filter symbol?) (set))
+                    args (maybe-type-hint-destructured-args clj? non-destructured-args args)
+                    {inner-ret-spec :spec
+                     ret-type-ann   :type-annotation} (-> args
+                                                          meta
+                                                          (extract-specs-from-metadata clj?)
+                                                          (first))
+                    args-sigs (map (fn [arg arg-meta]
+                                     (merge {:arg arg}
+                                            (-> arg-meta
+                                                (extract-specs-from-metadata clj?)
+                                                (first))))
+                                   args
+                                   (map meta args))
+                    args (cond-> (type-hint args args-sigs)
+                           (type-hint? ret-type-ann clj?) (vary-meta assoc :tag ret-type-ann))
+                    args-check-form (->> args-sigs
+                                         (concat specs-from-destructurings)
+                                         (filter :spec)
+                                         (map (fn [{:keys [spec arg]}]
+                                                {:pre [spec arg]}
+                                                [spec
+                                                 ;; Avoid "Can't type hint a primitive local" error:
+                                                 (vary-meta arg dissoc :tag)]))
+                                         (apply concat)
+                                         (vec))
+                    args-check-form? (-> args-check-form count pos?)
+                    ;; XXX analyze symbols (specs), replace keys (?)
```

Comment:
(this comment refers to the utils.spec checking possible API, and its possible differences to s.def's metadata API)